File: src/tangram/io.py
"""Load and save Tangram configurations as JSON, preserving exact Z[sqrt(2)] values."""
from __future__ import annotations

import json
from fractions import Fraction
from pathlib import Path

from .algebra import Z2
from .geometry import Point
from .model import PiecePlacement, Tangram
from .pieces import PieceType


def _fraction_to_json(value: Fraction):
    return value.numerator if value.denominator == 1 else str(value)


def _fraction_from_json(value) -> Fraction:
    return Fraction(value)


def _z2_to_json(z: Z2) -> list:
    return [_fraction_to_json(z.a), _fraction_to_json(z.b)]


def _z2_from_json(data: list) -> Z2:
    a, b = data
    return Z2.of(_fraction_from_json(a), _fraction_from_json(b))


def _piece_to_dict(piece: PiecePlacement) -> dict:
    return {
        "piece": piece.piece_type.value,
        "id": piece.piece_id,
        "anchor": {
            "x": _z2_to_json(piece.anchor.x),
            "y": _z2_to_json(piece.anchor.y),
        },
        "orientation": piece.orientation,
        "flipped": piece.flipped,
    }


def _piece_from_dict(data: dict) -> PiecePlacement:
    anchor = Point(_z2_from_json(data["anchor"]["x"]), _z2_from_json(data["anchor"]["y"]))
    return PiecePlacement(
        piece_type=PieceType(data["piece"]),
        piece_id=data["id"],
        anchor=anchor,
        orientation=data.get("orientation", 0),
        flipped=data.get("flipped", False),
    )


def tangram_to_dict(tangram: Tangram) -> dict:
    d: dict = {
        "name": tangram.name,
        "description": tangram.description,
        "pieces": [_piece_to_dict(p) for p in tangram.pieces],
    }
    if tangram.source:
        d["source"] = tangram.source
    return d


def tangram_from_dict(data: dict) -> Tangram:
    return Tangram(
        name=data["name"],
        description=data.get("description", ""),
        source=data.get("source", ""),
        pieces=[_piece_from_dict(p) for p in data["pieces"]],
    )


def save_tangram(tangram: Tangram, path: str | Path) -> None:
    Path(path).write_text(json.dumps(tangram_to_dict(tangram), indent=2) + "\n")


def load_tangram(path: str | Path) -> Tangram:
    return tangram_from_dict(json.loads(Path(path).read_text()))

File: web/src/io.ts
/**
 * Parse/serialize Tangram configurations to/from the same JSON schema used by
 * the Python project's io.py, preserving exact Z[sqrt(2)] values.
 */
import { Z2 } from "./algebra";
import { Fraction } from "./fraction";
import { Point } from "./geometry";
import { PiecePlacement, Tangram } from "./model";
import { PieceType } from "./pieces";

type JsonNum = number | string;

interface PieceJson {
  piece: string;
  id: number;
  anchor: { x: [JsonNum, JsonNum]; y: [JsonNum, JsonNum] };
  orientation: number;
  flipped: boolean;
}

interface TangramJson {
  name: string;
  description?: string;
  source?: string;
  pieces: PieceJson[];
}

function z2FromJson([a, b]: [JsonNum, JsonNum]): Z2 {
  return new Z2(Fraction.fromJSON(a), Fraction.fromJSON(b));
}

function z2ToJson(z: Z2): [JsonNum, JsonNum] {
  return [z.a.toJSON(), z.b.toJSON()];
}

function pieceFromJson(data: PieceJson): PiecePlacement {
  const anchor = new Point(z2FromJson(data.anchor.x), z2FromJson(data.anchor.y));
  return new PiecePlacement(
    data.piece as PieceType,
    data.id,
    anchor,
    data.orientation ?? 0,
    data.flipped ?? false,
  );
}

function pieceToJson(piece: PiecePlacement): PieceJson {
  return {
    piece: piece.pieceType,
    id: piece.pieceId,
    anchor: { x: z2ToJson(piece.anchor.x), y: z2ToJson(piece.anchor.y) },
    orientation: piece.orientation,
    flipped: piece.flipped,
  };
}

export function tangramFromJson(data: TangramJson): Tangram {
  return new Tangram(data.name, data.pieces.map(pieceFromJson), data.description ?? "", data.source ?? "");
}

export function tangramToJson(tangram: Tangram): TangramJson {
  const obj: TangramJson = {
    name: tangram.name,
    description: tangram.description,
    pieces: tangram.pieces.map(pieceToJson),
  };
  if (tangram.source) obj.source = tangram.source;
  return obj;
}

export async function loadTangram(url: string): Promise<Tangram> {
  const response = await fetch(url);
  if (!response.ok) throw new Error(`Failed to load ${url}: ${response.status}`);
  return tangramFromJson(await response.json());
}

export interface IndexEntry {
  file: string;
  title?: string;
  category: string;
  source: string;
  tags: string[];
}

export async function loadIndex(url: string): Promise<IndexEntry[]> {
  const response = await fetch(url);
  if (!response.ok) throw new Error(`Failed to load ${url}: ${response.status}`);
  const data = await response.json();
  return data.figures as IndexEntry[];
}
